File: secretletter/secretletter.go
package secretletter

import "strings"

type Decryptor interface {
	Decrypt(encrypted string) string
}

type decryptorImpl struct {
	shift int
}

func New(shift int) Decryptor {
	return &decryptorImpl{shift}
}

func (d *decryptorImpl) Decrypt(encrypted string) string {
	decryptedBytes := []byte{}
	for i := 0; i < len(encrypted); i++ {
		if i%2 == 0 {
			decryptedBytes = append(decryptedBytes, d.backward(encrypted[i]))
		} else {
			decryptedBytes = append(decryptedBytes, d.forward(encrypted[i]))
		}
	}
	return string(decryptedBytes)
}

const alphas = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

func (d *decryptorImpl) forward(b byte) byte {
	idx := strings.Index(alphas, string(b))
	idx = (idx + d.shift) % 26
	return alphas[idx]
}

func (d *decryptorImpl) backward(b byte) byte {
	idx := strings.Index(alphas, string(b))
	idx -= d.shift
	if idx < 0 {
		idx += 26
	}
	return alphas[idx]
}

File: src/secret_letter.rs
trait Decryptor {
    fn decrypt(&self, encrypted:&str) -> String;
}

fn new(shift:u32) -> impl Decryptor {
    DecryptorImpl::new(shift)
}

struct DecryptorImpl {
    shift: u32,
}

impl DecryptorImpl {
    fn new(shift:u32) -> DecryptorImpl{
        DecryptorImpl{shift}
    }

    fn forward(&self, ch: char) -> char {
        'A'
    }

    fn backward(&self, ch:char) -> char {
        'B'
    }
}

impl Decryptor for DecryptorImpl {
    fn decrypt(&self, encrypted:&str) -> String{
        let result = String::new();
        for i in 0..encrypted.len() {
            if i % 2 == 0 {
                result += self.backward(encrypted[i]);
            } else {
                result += self.forward(encrypted[i]);
            }
        }
        result
    }
}

#[cfg(test)]
mod tets {
    use super::Decryptor;
    #[test]
    fn test0(){
        let decryptor = super::new(4);
        let want = "MILK";
        let got = decryptor.decrypt("QEPG");
        assert_eq!(want, got);
    }

    #[test]
    fn test1(){
        let decryptor = super::new(19);
        let want = "RQRITJCB";
        let got = decryptor.decrypt("KXKPMQVI");
        assert_eq!(want, got);
    }
}